idx/memory/memory.go

```diff
@@ -465,6 +467,214 @@ func (m *MemoryIdx) TagDetails(orgId int, key, filter string, from int64) (map[s
+		if len(tagPrefix) > 0 {
+			expressions = append(expressions, fmt.Sprintf("__tag^=%s", tagPrefix))
+		}
+		query, err := NewTagQuery(expressions, from)
+		if err != nil {
+			return nil, err
+		}
+
+		// only acquire lock after we're sure the query is valid
+		m.RLock()
+		defer m.RUnlock()
+
+		tags, ok := m.tags[orgId]
+		if !ok {
+			return nil, nil
+		}
+
+		resMap := query.RunGetTags(tags, m.DefById)
+		for tag := range resMap {
+			res = append(res, tag)
+		}
+
+		// special case for the tag "name", we need to fake it
+		if len(res) > 0 && (len(tagPrefix) >= 4 && tagPrefix[:4] == "name") || len(tagPrefix) == 0 {
+			res = append(res, "name")
+		}
+
+		sort.Strings(res)
+		if uint(len(res)) > limit {
+			res = res[:limit]
+		}
+	} else {
+		m.RLock()
+		defer m.RUnlock()
+
+		tags, ok := m.tags[orgId]
+		if !ok {
+			return nil, nil
+		}
+
+		tagsSorted := make([]string, 0, len(tags))
+		for tag := range tags {
+			if len(tagPrefix) > 0 && (len(tagPrefix) > len(tag) || tag[:len(tagPrefix)] != tagPrefix) {
+				continue
+			}
+
+			tagsSorted = append(tagsSorted, tag)
+		}
+
+		sort.Strings(tagsSorted)
+
+	TAGS:
+		for _, tag := range tagsSorted {
+			for _, ids := range tags[tag] {
+				for id := range ids {
+					// only if from is specified we need to find at least one
+					// metric with LastUpdateTime >= from
+					if from > 0 {
+						def, ok := m.DefById[id.String()]
+						if !ok {
+							corruptIndex.Inc()
+							log.Error(3, "memory-idx: corrupt. ID %q is in tag index but not in the byId lookup table", id.String())
+							continue
+						}
+						if def.LastUpdate < from {
+							continue
+						}
+					}
+					res = append(res, tag)
+					if uint(len(res)) >= limit {
+						break TAGS
+					}
+					continue TAGS
+				}
+			}
+		}
+	}
+
+	return res, nil
+}
+
+// AutoCompleteTagValues is used to generate a list of possible values that could
+// complete a given value prefix. It requires a tag to be specified and only values
+// of the given tag will be returned. It also accepts additional conditions to
+// further narrow down the result set.
+//
+// tag:         values of which tag should be returned
+// valPrefix:   the string to be completed
+// expressions: tagdb expressions in the same format as graphite uses
+// from:        only tags will be returned that have at least one metric
+//              with a LastUpdateTime above from
+// limit:       the maximum number of results to return, the results will
+//              always be sorted alphabetically for consistency between
+//              consecutive queries
+//
+func (m *MemoryIdx) AutoCompleteTagValues(orgId int, tag, valPrefix string, expressions []string, from int64, limit uint) ([]string, error) {
+	var res []string
+
+	// only if expressions are specified we need to build a tag query.
+	// otherwise, the generation of the result set is much simpler
+	if len(expressions) > 0 {
+		if len(valPrefix) > 0 {
+			expressions = append(expressions, tag+"^="+valPrefix)
```

Comment:
Should this be "=~"?

Reply:
That's a special operator that MT knows about, just used for this stuff right now but I might add it to graphite also

Reply:
Ah, gotcha. Well, this doesn't work on my local build, but changing it to "=~" does. I'll figure that out real quick.